rir/src/interpreter/interp.c

```diff
@@ -1910,40 +1910,212 @@ INSTRUCTION(sub_) {
+                if (*INTEGER(lhs) == NA_INTEGER ||                             \
+                    *INTEGER(rhs) == NA_INTEGER) {                             \
+                    res = R_LogicalNAValue;                                    \
+                } else {                                                       \
+                    res = *INTEGER(lhs) op *INTEGER(rhs) ? R_TrueValue         \
+                                                         : R_FalseValue;       \
+                }                                                              \
+                break;                                                         \
+            }                                                                  \
+        }                                                                      \
+        BINOP_FALLBACK(#op);                                                   \
+    } while (false)
+
+
 INSTRUCTION(lt_) {
     SEXP lhs = ostack_at(ctx, 1);
     SEXP rhs = ostack_at(ctx, 0);
     SEXP res;
 
-    do {
-        if (IS_SCALAR_VALUE(lhs, REALSXP)) {
-            if (IS_SCALAR_VALUE(rhs, REALSXP)) {
-                if (*REAL(lhs) == NA_REAL || *REAL(rhs) == NA_REAL) {
-                    res = R_LogicalNAValue;
-                } else {
-                    res = *REAL(lhs) < *REAL(rhs) ? R_TrueValue : R_FalseValue;
-                }
-                break;
-            }
-        } else if (IS_SCALAR_VALUE(lhs, INTSXP)) {
-            if (IS_SCALAR_VALUE(rhs, INTSXP)) {
-                if (*INTEGER(lhs) == NA_INTEGER ||
-                    *INTEGER(rhs) == NA_INTEGER) {
-                    res = R_LogicalNAValue;
-                } else {
-                    res = *INTEGER(lhs) < *INTEGER(rhs) ? R_TrueValue
-                                                        : R_FalseValue;
-                }
-                break;
-            }
-        }
-        BINOP_FALLBACK("<");
-    } while (false);
+    DO_RELOP(<);
 
     ostack_popn(ctx, 2);
     ostack_push(ctx, res);
 }
 
+INSTRUCTION(gt_) {
+    SEXP lhs = ostack_at(ctx, 1);
+    SEXP rhs = ostack_at(ctx, 0);
+    SEXP res;
+
+    DO_RELOP(>);
+
+    ostack_popn(ctx, 2);
+    ostack_push(ctx, res);
+}
+
+INSTRUCTION(le_) {
+    SEXP lhs = ostack_at(ctx, 1);
+    SEXP rhs = ostack_at(ctx, 0);
+    SEXP res;
+
+    DO_RELOP(<=);
+
+    ostack_popn(ctx, 2);
+    ostack_push(ctx, res);
+}
+
+INSTRUCTION(ge_) {
+    SEXP lhs = ostack_at(ctx, 1);
+    SEXP rhs = ostack_at(ctx, 0);
+    SEXP res;
+
+    DO_RELOP(>=);
+
+    ostack_popn(ctx, 2);
+    ostack_push(ctx, res);
+}
+
+INSTRUCTION(eq_) {
+    SEXP lhs = ostack_at(ctx, 1);
+    SEXP rhs = ostack_at(ctx, 0);
+    SEXP res;
+
+    DO_RELOP(==);
+
+    ostack_popn(ctx, 2);
+    ostack_push(ctx, res);
+}
+
+INSTRUCTION(ne_) {
+    SEXP lhs = ostack_at(ctx, 1);
+    SEXP rhs = ostack_at(ctx, 0);
+    SEXP res;
+
+    DO_RELOP(!=);
+
+    ostack_popn(ctx, 2);
+    ostack_push(ctx, res);
+}
+
+INSTRUCTION(not_) {
```

Comment:
wouldn't you expect booleans to be common here? I think there should be in the fast path here and maybe also for others like `==`